File: wtf/validate-subsequence/validate-subsequence.js
// O(n) time | O(1) space
function validateSubsequenceV0(array, sequence) {
    let arrIdx = 0;
    let seqIdx = 0;

    while (arrIdx < array.length && seqIdx < sequence.length) {
        if (array[arrIdx] === sequence[seqIdx]) {
            seqIdx += 1;
        }

        arrIdx += 1;
    }

    return seqIdx === sequence.length;
}

// O(n) time | O(1) space
function validateSubsequenceV1(array, sequence) {
    let seqIdx = 0;

    for (let i = 0; i < array.length; i++) {
        if (seqIdx === sequence.length) {
            break;
        }

        if (array[i] === sequence[seqIdx]) {
            seqIdx += 1;
        }
    }

    return seqIdx === sequence.length;
}

const array = [4, 2, -1, 8, 0, 5, 9, 7];
const sequence = [2, -1, 0, 9];

console.log(validateSubsequenceV0(array, sequence));
console.log(validateSubsequenceV1(array, sequence));

File: wtf/validate-subsequence/validate-subsequence.py
# O(n) time | O(1) space
def validateSubsequenceV0(array, sequence):
    arrIdx = 0
    seqIdx = 0

    while arrIdx < len(array) and seqIdx < len(sequence):
        if array[arrIdx] == sequence[seqIdx]:
            seqIdx += 1

        arrIdx += 1

    return seqIdx == len(sequence)


# O(n) time | O(1) space
def validateSubsequenceV1(array, sequence):
    seqIdx = 0

    for value in array:
        if seqIdx == len(sequence):
            break
        if sequence[seqIdx] == value:
            seqIdx += 1

    return seqIdx == len(sequence)


array = [4, 2, -1, 8, 0, 5, 9, 7]
sequence = [2, -1, 0, 9]

print(validateSubsequenceV0(array, sequence))
print(validateSubsequenceV1(array, sequence))
